---
title: "WCB RNA-Seq Workshop"
author:  "WCB Bioinformatics Core Facility"
output:
  html_document:
    fig_width: 1
    df_print: paged
    css: "style.css"
---


<head>
<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>
</head>


```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

----

<br>

![](images/CELL_BIOLOGY_RGB.png){ width=50% }

<br>

This workshop is designed as an introduction to RNA-seq data analysis and will guide you through the following steps:

### RNA-seq analysis on the command line

* Assessing raw sequence data and performing quality control
* Aligning sequence reads to a reference genome
* Filtering alignments for further analysis
* Assessing the quality of your RNA-seq experiment
* Visualising alignments on a genome browser
* Quantifying expression levels

### Automated pipelines with NextFlow

* Install the nf-core RNA-seq pipeline
* Use singularity containers in your pipeline
* Run the pipeline with NextFlow

### Differential expression analysis with DESeq2

* Load transcript abundance files into R
* Run DESeq2 and visualise results
* Extract differentially expressed genes
* Perform functional enrichment on sets of genes

There is no such thing as a default pipeline. Although we mostly use standard parameters in this tutorial we hope to make you aware of the considerations you should take at each step. Make sure you understand your data and where it has come from. Use the correct tools for your dataset and read the tool documentation to see how different parameters affect your output!

This workshop is partly modeled on the training materials already provided by the [Harvard Chan Bioinformatics Core](https://hbctraining.github.io/main/). Please check out their website for comprehensive training materials in RNA-seq, NGS analysis and bioinformatics.

----

For more information contact [Shaun Webb](mailto:shaun.webb@ed.ac.uk).